# Autograd

In [1]:
import torch as t
import numpy as np
from torch.autograd import Variable

## 从Tensor建立

* 设定requires_grad为True，即可自动微分

In [2]:
x = t.ones(2, 2, requires_grad = True)
x

tensor([[1., 1.],
        [1., 1.]], requires_grad=True)

* 运算后的变量就会拥有grad_fn

In [3]:
y = x + 2
y

tensor([[3., 3.],
        [3., 3.]], grad_fn=<AddBackward0>)

In [4]:
y.grad_fn

* 更复杂的操作

In [5]:
z = y * y *3
out = z.mean()
print(z)
print(out)

tensor([[27., 27.],
        [27., 27.]], grad_fn=<MulBackward0>)
tensor(27., grad_fn=<MeanBackward0>)


* 使用 .requires_grad_( ... )来改变Tensor的require_grad的标志

In [6]:
a = t.randn(2, 2)
a = ((a * 3) / (a - 1))
print(a.requires_grad)
a.requires_grad_(True)
print(a.requires_grad)
b = (a * a).sum()
print(b.grad_fn)

False
True


## 梯度

* 进行反向传播

In [7]:
out.backward() # out只有一个值所以等于 out.backward(t.tensor(1.))

* 打印 d(out)/dx 的值

In [8]:
print(x.grad)

tensor([[4.5000, 4.5000],
        [4.5000, 4.5000]])


$out = \frac{1}{4}\sum_{i}{z_i}$ ， 而$z_i = 3(x+2)^2$ ，所以$\frac{\partial (out)}{\partial x_i}=\frac{\partial (out)}{\partial z_i}\frac{\partial (z_i)}{\partial x_i}=\frac{1}{4}*{6(x_i+2)}=\frac{3}{2}(x_i+2)$，因为$x_i=1$，所以结果是值均为$4.5$的$2*2$矩阵

$\vec{y}=f(\vec{x})$，
向量对向量计算梯度，本质上就是计算雅可比矩阵
$$J=\left(\begin{array}{ccc}
 \frac{\partial y_{1}}{\partial x_{1}} & \cdots & \frac{\partial y_{1}}{\partial x_{n}}\\
 \vdots & \ddots & \vdots\\
 \frac{\partial y_{m}}{\partial x_{1}} & \cdots & \frac{\partial y_{m}}{\partial x_{n}}
 \end{array}\right)$$